postgresql_lexer_base.go

```diff
@@ -51,6 +51,47 @@ func (receiver *PostgreSQLLexerBase) CheckLaMinus() bool {
 	return receiver.GetInputStream().LA(1) != '-'
 }
 
+/**
+ * Checks if the next character is a template parameter, not an operator.
+ * This will force the tokenizer to split the tokens instead of making it one long operator.
+ */
+func (receiver *PostgreSQLLexerBase) CheckLaTemplateParam() bool {
+	// Handle the case where we have "=?"
+	if receiver.isNextCharacterTemplateParam() {
+		return true
+	}
+
+	// Handle case where we've already parsed the template param because it's on the left of an operator (e.g. "?=")
+	return receiver.isCurrentCharacterTemplateParam()
+}
+
+func (receiver *PostgreSQLLexerBase) isNextCharacterTemplateParam() bool {
+	if receiver.GetInputStream().LA(1) != '?' {
+		return false
+	}
+
+	text := receiver.GetText()
+	if len(text) == 0 {
+		return false
+	}
+
+	// My assumption is that it only makes sense to have a template param after a comparison operator
```

Comment:
is this assumption necessary? thinking about functions as a counterexample. `SELECT * FROM abc WHERE def = UPPER(?);`

Reply:
Since this is inside the `Operator` rule in the lexer, the function and the comment are only in the context of matching an `Operator` token.

The lexer rule tries to group together multiple adjacent operator characters into a single operator character so that something like `>=` is a single operator, not two side-by-side operator.

So the assumption is that if a `?` is next to another operator character and that character is a comparison operator, then the `?` is a template param.

The example you've provided is a good edge case, and I've added a test for it to the bitbucket PR. It passes with the current implementation because it's a single ? alone and not next to a comparison operator.